packages/apriltag/src/apriltag_detector_node.py

```diff
@@ -0,0 +1,185 @@
+
+
+class AprilTagDetector(DTROS):
+
+    def __init__(self):
+        super(AprilTagDetector, self).__init__(
+            node_name='apriltag_detector_node',
+            node_type=NodeType.PERCEPTION
+        )
+        # get static parameters
+        self.family = rospy.get_param('~family', 'tag36h11')
+        self.nthreads = rospy.get_param('~nthreads', 1)
+        self.quad_decimate = rospy.get_param('~quad_decimate', 1.0)
+        self.quad_sigma = rospy.get_param('~quad_sigma', 0.0)
+        self.refine_edges = rospy.get_param('~refine_edges', 1)
+        self.decode_sharpening = rospy.get_param('~decode_sharpening', 0.25)
+        self.tag_size = rospy.get_param('~tag_size', 0.065)
+        # dynamic parameter
+        self.detection_freq = DTParam(
+            '~detection_freq',
+            default=-1,
+            param_type=ParamType.INT,
+            min_value=-1,
+            max_value=30
+        )
+        self._detection_reminder = DTReminder(frequency=self.detection_freq.value)
+        # camera info
+        self._camera_parameters = None
+        self._camera_frame = None
+        # create detector object
+        self._at_detector = Detector(
+            families=self.family,
+            nthreads=self.nthreads,
+            quad_decimate=self.quad_decimate,
+            quad_sigma=self.quad_sigma,
+            refine_edges=self.refine_edges,
+            decode_sharpening=self.decode_sharpening
+        )
+        # create a CV bridge object
+        self._bridge = CvBridge()
+        # create subscribers
+        self._img_sub = rospy.Subscriber('image_rect', Image, self._img_cb)
+        self._cinfo_sub = rospy.Subscriber('camera_info', CameraInfo, self._cinfo_cb)
+        # create publishers
+        self._img_pub = rospy.Publisher(
+            'tag_detections_image/compressed', CompressedImage, queue_size=1,
+            dt_topic_type=TopicType.VISUALIZATION
+        )
+        self._tag_pub = rospy.Publisher(
+            'tag_detections', AprilTagDetectionArray, queue_size=1,
+            dt_topic_type=TopicType.PERCEPTION
+        )
+        self._img_pub_busy = False
+        # create TF broadcaster
+        self._tf_bcaster = tf.TransformBroadcaster()
+        # spin forever
+        rospy.spin()
+
+    def _img_cb(self, data):
+        if self._camera_parameters is None:
+            return
+        if not self._detection_reminder.is_time(frequency=self.detection_freq.value):
+            return
+        # ---
+        # turn image message into grayscale image
+        img = self._bridge.imgmsg_to_cv2(data, desired_encoding='mono8')
+        # detect tags
+        tags = self._at_detector.detect(img, True, self._camera_parameters, self.tag_size)
+        # draw the detections on an image (if needed)
+        if self._img_pub.anybody_listening() and not self._img_pub_busy:
+            self._img_pub_busy = True
+            Thread(target=self._publish_detections_image, args=(img, tags)).start()
```

Comment:
Why a separate thread? Isn't this dangerous if generating the image is too slow and we start threads faster than they exit?